doc/sync-label.md

```diff
@@ -0,0 +1,224 @@
+# Sync Label
+
+## Motivation
+
+I am currently using git-bug to manage my bug/feature/task list for a number
+of projects. Some of those projects have "external" bug trackers (e.g.
+my day job, or github mirrors of side projects).
+
+My personal bug repo covers 20 projects. I prefer the experience of
+interacting with a single git-bug session in `termui` versus a half-dozen
+different laggy-ass hyper-ajax web 2.0 online trackers.
+
+I would love to sync my personal bug repo with these external trackers but
+there are a couple of problems:
+
+1. Our JIRA at work has millions of bugs. I obviously don't interact with
+   all of them. I only want to sync bugs that are "relevant" to me... where
+   "relevant" can presumably be defined by matching some criteria on metadata
+   in JIRA.
+2. Our gitlab at work likewise has thousands of open pull requests. I'm not sure
+   git-bug is necessarily the best tool for doing code-review right now, but
+   at the very least it might be nice to trach which code-reviews that I'm
+   involved in are still open and require attention.
+3. Some of the bugs in my repo really shouldn't be pushed to a public location
+   because they might contain "sensitive" information about my employer.
+4. I don't want bugs for irrelevant projects to pollute the issue list on
+   my github mirrors.
+
+## Proposed Solution
+
+My proposed solution is to restict synchronization to bugs that match certain
+filters, and to modify bugs as they are imported/exported according to rules
+in the git config. Initially, the filter will support matching labels, and
+modifiers will support adding or removing labels.
+
+NOTE: I may also take a stab at implementing a JIRA bridge.
+
+### Case Study
+
+Consider the project `git-bug-test` which has a github mirror at
+`cheshirekow.com/git-bug-test`. Ideally we would specify in our git-config
+something like this:
+
+```config
+[git-bug "bridge.github.github-git-bug-test"]
+	user = cheshirekow
+	project = git-bug-test
+	token = xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+	import-filter="assignee=cheshirekow"
+	import-mod="label=+test,label=+github"
+	export-filter="labels=test&github"
+	export-mod="label=-test,label=-github"
+```
+
+with the following semantics:
+
+**import**: Only issues matching `assignee=cheshirekow` will be queried
+from github. Any issue that is created or updated is given the labels `test`
+and `github` in `git-bug`.
+
+**export**: Only bugs which have *both* labels `test` and `github` will be
+considered for export. When we export the bugs, we remove the labels `test`
+and `github`. Or, rather, we do not synchronize those labels with github.
+
+### Open questions:
+
+To what extent can this implementation be generic between different remote
+bug systems? Each remote system will have different sets of metadata that we
+may or may not be able to filter on.
+
+Should we validate that `import-mod` matches `export-filter`? If not then
```

Comment:
I think we generally agree on the requirement. The actual way to implement that is less clear though. There is some important UX consequences here so we should take the time to get that right.

> A specific project piece of metadata would be nice and would enable the same features I'm looking for.

Isn't that essentially the same thing as the origin of the bug (bugtracker type + project) ?

> I still think modifying data on import would be very useful. Perhaps modifying it again on export isn't necessary that metadata isn't being overloaded for exportability tagging.

My hope is that using the metadata system is enough, without additional user-controller tagging. This way, the labelling keep being its own thing, there is no need to cleanup anything before exporting, metadata being by nature a purely git-bug thing. They are stored with the bug though, which means they get pushed and pulled through a git remote. If there is no need for user-custom metadata this also stay its own thing.

> Maybe a bug can be "assigned to" more than one "remote". That would be cool.

Yes, possibly, but I would wait for that to be obviously necessary.

> If nothing else, you could then use git-bug to bridge two bug systems.

Maybe, but there is the limitation at the moment that a bridge can't export something with an identity (=user account) you don't have a token for. So you can't import everything from a bugtracker and push to another one.

Reply:
Also a note: `Query` are not aware of metadata at the moment so it will be needed to expand that if that's the way to go.

Reply:
> Isn't that essentially the same thing as the origin of the bug (bugtracker type + project) ?

I wasn't originally thinking that "project" and "bugtracker type" would be a one-to-one mapping, but if you're advocating for that, I think it works fine for my needs. If ever there was a reason for "multi-remote" tagging that might be satisfied by "multi-project" assignment... so the idea seems to be scalable on the surface. It's maybe worth thinking about a couple of places where this model might not fit well and two come to mind:

1. monorepo (code) composed of multiple projects
2. mono-bugtracker composed of multiple projects - i.e. JIRA for a medium to large sized company

I don't think either of these usecases prevents the project:bucktracker 1-1 mapping though. It might just mean that you need to configure two (or more) bridges that point to the same remote bugtracker (one per "project") which I am fine with.

In any case, it certainly works for me. Although... 

> My hope is that using the metadata system is enough, without additional user-controller tagging.

My preference would be that anything I create in `git-bug` would not "export" automatically, unless I ask it to. I would, however, still like to associate it with a particular project so that I can see all bugs relevant to that project in the termui and such, so I think I would still like it to have a "project assignment". 

To be a little more concrete, let's talk specifically about "Project X" and let's assume that we have `project` as a new field in `git-bug` (or at least exposed through the UI, possibly implemented using existing metadata). Futhermore, let's assume that I have a "Project X" mirror on github with "issues" enabled so that users can report issues, but that `git-bug` is the "canonical" bug tracker. If a user of Project X reports a bug on github, I would like that bug to be imported into git-bug and assigned the `project=X` metadata. That way I can find it with all my other `project=X` bugs. However, I don't necessarily want all the git-bugs with the `project=X` label to be exported back to github. 

I can respect that maybe others *would* want them to automatically go to github, but I would prefer to explicitly choose. 

That said, I'm not fixated on using labels for this purpose. If `project` were available as a piece of metadata, maybe what I'm asking for is an addition boolean flag called `public=true`, or maybe `private=false` or maybe `export=true` (by default, whereas for my preference I would config it to be the opposite). I guess the semantics would be: Any bug whose origin is the remote tracker is `private=false`, any bug whose origin is `git-bug` is `private=<user-config,default=false>` initially. Then add some UI to flip the flag.

Reply:
It make sense to me to be able to control if you want to export "free floating" bugs. However as the majority of users will have one repo for one external bugtracker, it make sense to keep exporting everything as the default.

> It might just mean that you need to configure two (or more) bridges that point to the same remote bugtracker (one per "project") which I am fine with.

At the moment, bridges will tag bugs with the target (github, gitlab ..) and another specific information (the project URL for github). So having multiple bridges for the same bugtracker won't help you.

However, in the case of monorepo or mono-bugtracker, I expect either:
- remote bugs have a project specific label (that is, a user specific way to classify them), so you can filter based on that
- bugs are in some sort of global pool with a bugtracker specific way to classify them (a project field or something), so the bridge implementation could use that to define what is a project. This way, you could have multiple bridges with the same remote bugtracker, one for each project you care about. Bugs would be tagged distincltively.

Reply:
> However, in the case of monorepo or mono-bugtracker, I expect either: ...

Think this is also what I was thinking.